docs/design/persistent-storage.md

```diff
@@ -0,0 +1,222 @@
+
+
+
+#### Pod to specify storage requirements
+
+```
+
+	Pod.Spec.Volumes.Volume 
+			Name
+			VolumeSource -- to be bound with available storage from the pool
+			PersistenceRequirements
+				Mode				ReadWriteOnce | ReadOnlyMany
+				RetentionPolicy 	StoreRetentionPolicy (RetainOnDelete|RecycleOnDelete)				Resources			ResourcesList (size, iops, throughput)
+				Selector			Labels to identify the PersistentStore backing implementation
+
+```
+
+`PersistenceRequirements` are the capabilities and performance characteristics a pod author can request and rely on for the entire lifecycle of the volume.  Requests for storage that do not match available storage will go unfulfilled.
+
+
+#### Storage discovery 
+
+An API and CLI is needed to view available and in-use stores by namespace.
+
+**Use Case**:  Susie needs to crunch some data and serve it afterwards.  A single pod to create data is sufficient, but she estimates needing 10 web front ends to serve the data afterwards.  For this use case, the data does not change once created.  Susie first looks to see what is available in the cluster.
+
+```
+$ cluster/kubectl.sh get storage
+
+NAME                LABELS              MODES               RESOURCES
+_pv_vol_01          name=my-data        RWO,ROX             size=5gb,iops=30
+_pv_vol_02                              RWO,ROX             size=5gb,iops=30
+_pv_vol_03          name=other-data     RWO,ROX             size=25gb,iops=30
+
+```
+
+#### Request persistent storage for a pod
+
+Susie is happy to see storage that can be mounted many times (ReadOnlyMany).  That satisfies her load balancer requirement.  Susie creates an pod image with her number crunching code and requests persistence for her volume.
+
+Susie's YAML looks something like this for the pod that creates her data:
+
+
+```yaml
+
+id: creator-of-data
+kind: Pod
+namespace: ns
+apiVersion: v1beta1
+desiredState:
+  manifest:
+    version: v1beta1
+    id: creator-of-data
+    containers:
+      - name: numbercruncher
+        image: ns/makesomdata:v1
+        command:
+          - /run.sh
+        volumeMounts:
+          - name: data
+            mountPath: /some_data
+    volumes:
+      - name: data
+        persistenceRequirements:
+        	availableModes: readwriteonce | readonlymany
+        	currentMode: readwriteonce
+        	retentionPolicy: retainondelete
+        	resources:
+        		- size: 5
+        	selector:
+        		- name: my-persistent-data
+labels:
+  name: creator-of-data
+
+```
+
+#### Behind the curtain
+
+`PersistentStoreBinder` is watching for new pods with `PersistenceRequirements`.  The binder compares the request to all available stores and binds it to the closest match.
+
+To bind:
+
+* Set PersistentStore.Status.PodRef = pod that originally requested persistence
+* Set PersistentStore.Labels = PersistenceRequirements.Selector
```

Comment:
`* Set PersistentStore.Labels = PersistenceRequirements.Selector`
I don't like that the labels were copied from the selector. Like with replication controller, I'd separate the labels in the template and the selector to match them.

Re. claims:
I'm not sure we want to distinguish new vs. not new. What if the Pod with volume NewPersistent doesn't fully initialize the volume? How would the PSController know to rebind it to a NewPersistent volume rather than a Persistent one? Initialization needs to be idempotent, or done very carefully manually, in which case I'm not sure the mechanism helps.

This model also isn't as easy to use with replication controllers as I'd like. Cloud-native self-replicating stores, like etcd or cassandra, aren't pets but would benefit from network volumes.

If we wanted to have a distinct "claim" type, analogous to a pod -- a logical use of the network volume, I'd make a controller manage them together with pods. It could create N volume claims for N pods. Whether this is a new controller or we extend replication controller is another issue. If a user wanted to initialize volumes, they'd have to spin up pods to do it, then change/add labels on the claims to enable them to match requests that required initialized volumes.

Note also, however, the discussion on #1515 and in #260 -- it's a pod's IP address AND its data that are significant to its identity. If we're going to support something like nominal services, identity/role assignment needs to take that into account. That's a much finer-grain notion than initialized vs. not.